### PR TITLE
Testing telegram message and sticker handlers

### LINKED ISSUE #272: Write unit tests for the message handler
# Summary

Unit tests do not currently exist for the message handler. 

# Background

**Is your improvement related to a problem? Please describe**:

After writing the message handler, it is unclear if we have handled all edge cases. We should have unit tests to thoroughly test the message handler functionality.

**Describe the solution you'd like**:

Full set of unit tests for the message handler to catch as many edge cases as we can.

**Describe alternatives you've considered**:

No alternatives.

# Details

The owner of this issue needs to think of what the message handler should support, and not support. Then, build unit tests around the desired and undesired functionality to make sure error handling exists in the appropriate locations.

# Outcome

Unit testing encompasses the message handler. Allows us to feel safe knowing the message handler has been thoroughly tested.

### LINKED ISSUE #287: Write unit tests for the sticker handler
# Summary

Unit tests do not currently exist for the Telegram sticker handler.

# Background

**Is your improvement related to a problem? Please describe**:

After writing this handler, it is unclear if we have handled all edge cases. We should have unit tests to thoroughly test this functionality.

**Describe the solution you'd like**:

Full set of unit tests to catch as many edge cases as we can.

**Describe alternatives you've considered**:

No alternatives.

# Details

The owner of this issue needs to think about what a user can and cannot do when using a sticker in Telegram. Then, build unit tests around the desired and undesired functionality to make sure error handling exists in the appropriate locations.

# Outcome

Unit testing for the TG sticker handler. Allows us to feel safe knowing this handler is thoroughly tested.

### PR DESCRIPTION
**Pull request by kennedy**
For each test per handler, I created a client and update object and pass that through. Is this what we are looking for?

resolves #272
resolves #287 

### passing
```
=== RUN   ExampleStickerHandler
--- PASS: ExampleStickerHandler (0.00s)
=== RUN   ExampleMessageHandler
--- PASS: ExampleMessageHandler (0.00s)
PASS

Process finished with exit code 0
```

### failing
```
=== RUN   ExampleMessageHandler
--- FAIL: ExampleMessageHandler (0.00s)
got:
<test>Random Text
want:
<test>Random wrong Text
FAIL

Process finished with exit code 1
```

### CODE REVIEW

**Comment by jwflory** [comment]
I defer to the higher wisdom of @Zedjones and @Tjzabel on this one.

**Comment by kennedy** [comment]
> 
> 
> You're gonna want something programmatic to verify that your output is correct rather than a print statement. Maybe use an assertion there instead? I haven't looked at the handler yet, these are just some preliminary comments.

@Zedjones @Tjzabel There seems to be different types of testing design by golang. The one I am following is the Example type https://golang.org/pkg/testing/#hdr-Examples

If you were to run go test against it, you will get the above results

**Comment by Tjzabel** [comment]
@kennedy so in this case, we want to use the default `TestStickerHandler(*testing.T)` type functions.

**Comment by kennedy** [comment]
@Tjzabel @Zedjones I converted to the generic test cases

File: internal/handlers/telegram/handler.go [hunk 1]
```diff
@@ -56,7 +57,7 @@ func stickerHandler(tg *Client, u tgbotapi.Update) {
 	formatted := tg.Settings.Prefix + u.Message.From.UserName +
 		tg.Settings.Suffix + u.Message.Sticker.Emoji
 
-	tg.sendToIrc(formatted)
+	tg.sendToIrc(strings.Trim(formatted," "))
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler.go:60]: What is this trimmed for?

**Comment by Tjzabel** [comment]
@kennedy looks good so far. Can you please fix the CodeClimate formatting error? I will take a more in-depth look with this before our next sprint meeting.

File: internal/handlers/telegram/handler.go (hunk 1 above)
Comment (kennedy) [reply internal/handlers/telegram/handler.go:60]: There was an extra space at the end of the string. It was not expected, so I'm trimming it

Comment (jwflory) [reply internal/handlers/telegram/handler.go]: @kennedy This would probably be a good place to leave a comment for context. We'll probably forget why we did this in another twelve months, maybe. :smile:

File: internal/handlers/telegram/handler.go [hunk 2]
```diff
@@ -56,7 +57,8 @@ func stickerHandler(tg *Client, u tgbotapi.Update) {
 	formatted := tg.Settings.Prefix + u.Message.From.UserName +
 		tg.Settings.Suffix + u.Message.Sticker.Emoji
 
-	tg.sendToIrc(formatted)
+	// Trim unexpected trailing whitespace
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler.go:60]: @kennedy this whitespace is actually purposely done. All messages start with a space after the ending suffix. Message should say
```
<test> Random Text
```
and _**not**_:
```
"<test>Random Text"
```

**Comment by Zedjones** [comment]
Personally, I think that the `assert` library achieves what you're trying to do in a bit of a cleaner way. Instead of this:
```go
if s != correct {
    t.Errorf(fmt.Sprintf("messageHandler(\"Random Text\") = \"%s\"; want \"%s\"", s, correct))
    assert.Equal(t, s, correct, "Testing")
}
```
You can just do this:
```go
assert.Equal(t, s, correct, "Some error message")
```
And then the output looks like this:
```
--- FAIL: TestMessageRandom (0.00s)
    hander_test.go:56: 
                Error Trace:    hander_test.go:56
                                                        handler.go:49
                                                        hander_test.go:60
                Error:          Not equal: 
                                expected: "<test>Random Text"
                                actual  : "<test>Random Textf"
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -<test>Random Text
                                +<test>Random Textf
                Test:           TestMessageRandom
```
Which is easier to read, in my opinion

File: internal/handlers/telegram/handler.go (hunk 2 above)
Comment (kennedy) [reply internal/handlers/telegram/handler.go:60]: According to the formatted string, it was "<test>Random Text ". The trim removes the trailing " " at the end

**Comment by kennedy** [comment]
> Which is easier to read, in my opinion

I agree this looks much better, and the formatting of errors comes for free. I updated the test to use the `github.com/stretchr/testify/assert` library

File: internal/handlers/telegram/hander_test.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestStickerSmile(t *testing.T) {
+	correct := "<test>😄"
+	updateObj := tgbotapi.Update{
+		Message: &tgbotapi.Message{
+			From: &tgbotapi.User{
+				UserName: "test",
```
Comment (Tjzabel) [inline internal/handlers/telegram/hander_test.go:16]: Side note, but the `UserName` field is not required for Telegram users, only the `Name` field. 
To get around that, use the `user.String()` method which grabs the username if present, or the name otherwise.

File: internal/handlers/telegram/hander_test.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+			},
+		},
+	}
+	clientObj := &Client{
+		Settings: internal.TelegramSettings{
+			Prefix: "<",
+			Suffix: ">",
+		},
+		sendToIrc: func(s string) {
+			assert.Equal(t, correct, s)
```
Comment (Tjzabel) [inline internal/handlers/telegram/hander_test.go:29]: :+1:

File: internal/handlers/telegram/hander_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
+	"github.com/kyokomi/emoji"
+	"github.com/ritlug/teleirc/internal"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestStickerSmile(t *testing.T) {
```
Comment (Tjzabel) [inline internal/handlers/telegram/hander_test.go:11]: So this is a good start, but this only tests if everything is working correctly. What if we have a sticker object, but there isn't an emoji present? What if the `UserName` isn't present?

File: internal/handlers/telegram/hander_test.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		},
+		sendToIrc: func(s string) {
+			assert.Equal(t, correct, s)
+		},
+	}
+
+	stickerHandler(clientObj, updateObj)
+}
+
+func TestMessageRandom(t *testing.T) {
```
Comment (Tjzabel) [inline internal/handlers/telegram/hander_test.go:36]: Same thing with sticker message. This is only testing our positive test cases. Same thing as above, what if `UserName` field is missing, etc.

File: internal/handlers/telegram/hander_test.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+	clientObj.Settings = internal.TelegramSettings{
+		Prefix: "<",
+		Suffix: ">",
+	}
+	clientObj.SendToIrc = func(s string) {
+		assert.Equal(t, correct, s)
+	}
+
+	// clientObj.handler.stickerHandler(clientObj, updateObj)
```
Comment (kennedy) [inline internal/handlers/telegram/hander_test.go:35]: I need some help. Turning the test file to telegram_test, though I agree this is the best course of action, I cant call the sticketHandler directly. 

What @Zedjones did with mocking the client in testing IRC is a better structure, so I will attempt to follow suite. However, his project isnt in the irc_test namespace either. 

Can someone help me get a better grasp of how to handle this?

File: internal/handlers/telegram/hander_test.go (hunk 3 above)
Comment (kennedy) [reply internal/handlers/telegram/hander_test.go:17]: According to how the messages handler is forming the formatted message, UserName was used directly. Where can I find the `user` object?
```
formatted := tg.Settings.Prefix + u.Message.From.UserName +
		tg.Settings.Suffix + u.Message.Text
```

File: internal/handlers/telegram/telegram.go [hunk 8]
```diff
@@ -14,7 +14,7 @@ type Client struct {
 	api       *tgbotapi.BotAPI
 	Settings  internal.TelegramSettings
 	logger    internal.DebugLogger
-	sendToIrc func(string)
+	SendToIrc func(string)
```
Comment (kennedy) [inline internal/handlers/telegram/telegram.go:17]: I had to capitalize the SendToIrc inorder to assign it outside the telegram namespace for testing. 

I dont think this is the proper way, but I dont know any other way

File: internal/handlers/telegram/hander_test.go (hunk 3 above)
Comment (kennedy) [reply internal/handlers/telegram/hander_test.go:17]: nvm, I figured out what you meant

File: internal/handlers/telegram/handler.go [hunk 9]
```diff
@@ -42,9 +44,12 @@ messageHandler handles the Message Telegram Object, which formats the
 Telegram update into a simple string for IRC.
 */
 func messageHandler(tg *Client, u tgbotapi.Update) {
-	formatted := tg.Settings.Prefix + u.Message.From.UserName +
-		tg.Settings.Suffix + u.Message.Text
-
+	formatted := fmt.Sprintf("%s%s%s %s",
```
Comment (kennedy) [inline internal/handlers/telegram/handler.go:47]: Added space between username and text

File: internal/handlers/telegram/handler.go (hunk 2 above)
Comment (kennedy) [reply internal/handlers/telegram/handler.go]: fixed

File: internal/handlers/telegram/hander_test.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
+package telegram
```
[the remark this answers is not in the document]
Comment (kennedy) [reply internal/handlers/telegram/hander_test.go]: fixed

File: internal/handlers/telegram/hander_test.go (hunk 3 above)
Comment (kennedy) [reply internal/handlers/telegram/hander_test.go]: fixed

File: internal/handlers/telegram/hander_test.go (hunk 6 above)
Comment (kennedy) [reply internal/handlers/telegram/hander_test.go]: fixed

File: internal/handlers/telegram/hander_test.go (hunk 5 above)
Comment (kennedy) [reply internal/handlers/telegram/hander_test.go]: How would there be a sticker with no emoji?

File: internal/handlers/telegram/handler.go [hunk 11]
```diff
@@ -22,8 +24,10 @@ func updateHandler(tg *Client, updates tgbotapi.UpdatesChannel) {
 			tg.logger.LogError("Missing message data")
 			continue
 		case u.Message.NewChatMembers != nil:
+			tg.logger.LogDebug("joinHandler triggered")
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler.go:27]: thanks for catching this :+1:

File: internal/handlers/telegram/handler.go [hunk 12]
```diff
@@ -47,9 +51,12 @@ messageHandler handles the Message Telegram Object, which formats the
 Telegram update into a simple string for IRC.
 */
 func messageHandler(tg *Client, u tgbotapi.Update) {
-	formatted := tg.Settings.Prefix + u.Message.From.UserName +
-		tg.Settings.Suffix + u.Message.Text
-
+	formatted := fmt.Sprintf("%s%s%s %s",
+		tg.Settings.Prefix,
+		GetUsername(u.Message.From),
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler.go:56]: So this is actually a bit of a misnomer. I thought about changing the name of this function to `GetUsernameFull` or something to that extent. This is just a helper function to properly format a person's name and username into
```
test (@testUser)
```
What _you_ want to do here is call `u.Message.From.String()` which will properly grab the `UserName` if exists, or `FirstName` field otherwise.

File: internal/handlers/telegram/handler.go [hunk 13]
```diff
@@ -83,9 +90,13 @@ stickerHandler handles the Message.Sticker Telegram Object, which formats the
 Telegram message into its base Emoji unicode character.
 */
 func stickerHandler(tg *Client, u tgbotapi.Update) {
-	formatted := tg.Settings.Prefix + u.Message.From.UserName +
-		tg.Settings.Suffix + u.Message.Sticker.Emoji
 
+	formatted := fmt.Sprintf("%s%s%s %s",
+		tg.Settings.Prefix,
+		GetUsername(u.Message.From),
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler.go:96]: Same as above. `u.Message.From.String()`

File: internal/handlers/telegram/handler.go [hunk 14]
```diff
@@ -83,9 +90,13 @@ stickerHandler handles the Message.Sticker Telegram Object, which formats the
...
+		tg.Settings.Suffix,
+		// Trim unexpected trailing whitespace
+		strings.Trim(u.Message.Sticker.Emoji, " "))
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler.go:99]: Honestly just not sure why there would even be a trailing whitespace char here, but doesn't hurt

File: internal/handlers/telegram/handler_test.go [hunk 15]
```diff
@@ -130,3 +132,122 @@ func TestJoinNoUsername(t *testing.T) {
...
 }
+
+func TestStickerSmile_withusername(t *testing.T) {
+	testUser := &tgbotapi.User{
+		ID:        1,
+		UserName:  "test",
+		FirstName: "testing",
+		LastName:  "123",
+	}
+	correct := fmt.Sprintf("<%s (@%s)> 😄", testUser.FirstName, testUser.UserName)
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler_test.go:143]: I would prefer we keep to only grabbing the username when sending messages in their name.
`<fmt.Sprintf("<%s> 😄", testUser.UserName)`

File: internal/handlers/telegram/handler_test.go [hunk 16]
```diff
@@ -130,3 +132,122 @@ func TestJoinNoUsername(t *testing.T) {
 	}
 	joinHandler(clientObj, testListUser)
 }
+
+func TestStickerSmile_withusername(t *testing.T) {
```
Comment (Tjzabel) [inline internal/handlers/telegram/handler_test.go:136]: I would like to only use CamelCase in unit tests.

File: internal/handlers/telegram/handler.go (hunk 12 above)
Comment (kennedy) [reply internal/handlers/telegram/handler.go:56]: alright Ill update it

File: internal/handlers/telegram/handler.go (hunk 14 above)
Comment (kennedy) [reply internal/handlers/telegram/handler.go:99]: I think its the library I am using in testing to convert strings to emoji. They must have add a space at the end

File: internal/handlers/telegram/handler_test.go (hunk 16 above)
Comment (kennedy) [reply internal/handlers/telegram/handler_test.go]: fixed

File: internal/handlers/telegram/handler_test.go (hunk 15 above)
Comment (kennedy) [reply internal/handlers/telegram/handler_test.go]: fixed

File: internal/handlers/telegram/handler.go (hunk 14 above)
Comment (kennedy) [reply internal/handlers/telegram/handler.go]: fixed

File: internal/handlers/telegram/handler.go (hunk 13 above)
Comment (kennedy) [reply internal/handlers/telegram/handler.go]: fixed

File: internal/handlers/telegram/handler.go (hunk 12 above)
Comment (kennedy) [reply internal/handlers/telegram/handler.go]: fixed

**Comment by jwflory** [comment]
Woot! Nice to see this get merged. Thanks @kennedy :tada:
